hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -239,16 +285,20 @@ def set_element(self, val, result, idx, *, fill=False):
             result[idx] = val
         except TypeError as err:
             raise InvalidArgument(
-                f"Could not add element={val!r} of {val.dtype!r} to array of "
-                f"{result.dtype!r} - possible mismatch of time units in dtypes?"
+                f"Could not add element={getattr(val, 'dtype', type(val))!r} of "
+                f"{getattr(val, 'dtype', type(val))!r} to array of "
+                f"{getattr(result, 'dtype', type(result))!r} - possible mismatch of time units in dtypes?"
             ) from err
         try:
-            elem_changed = self._check_elements and val != result[idx] and val == val
+            elem_changed = self._check_elements and (
+                not _array_or_scalar_equal(val, result[idx])
+                and _array_or_scalar_equal(val, val)
+            )
```

Comment:
I'm concerned that this will be more expensive in a pretty hot loop - can we check that it doesn't slow down e.g. arrays or floats, or perhaps do this more complicated check only for object arrays?